### PR TITLE
[16.0][MIG] stock_request_picking_type: Migration to 16.0

### PR DESCRIPTION
**Pull request by peluko00**
Module migrated to version 16.0

cc https://github.com/APSL 153677

@miquelalzanillas @lbarry-apsl @javierobcn @mpascuall please review

### CODE REVIEW

**Review by mpascuall — APPROVED** [review]
Works correctly, checked in runboat

File: stock_request_picking_type/tests/test_stock_request.py [hunk 1]
```diff
@@ -34,47 +34,35 @@ def setUpClass(cls):
...
-                    "company_ids": [(6, 0, company_ids)],
+                    "login": "stock_request_user",
+                    "email": "@".join(["stock_request_user", "test.com"]),
+                    "groups_id": [(6, 0, [cls.stock_request_user_group.id])],
+                    "company_ids": [(6, 0, [cls.main_company.id, cls.company_2.id])],
                 }
             )
         )
 
-    def _create_product(self, default_code, name, company_id, **vals):
```
Comment (rousseldenis) [inline stock_request_picking_type/tests/test_stock_request.py]: Don't remove those kind of helper methods.

File: stock_request_picking_type/tests/test_stock_picking_type.py [hunk 2]
```diff
@@ -74,39 +90,11 @@ def setUpClass(cls):
             "stock.no_auto_scheduler", "True"
         )
 
-    def _create_user(self, name, group_ids, company_ids):
```
Comment (rousseldenis) [inline stock_request_picking_type/tests/test_stock_picking_type.py]: Don't remove those helper methods.

**Review by rousseldenis — CHANGES_REQUESTED** [review]

**Comment by peluko00** [comment]
Fix it! Can review again please @rousseldenis

**Review by lbarry-apsl — APPROVED** [review]
Reviewed

File: stock_request_picking_type/tests/test_stock_picking_type.py (hunk 2 above)
Comment (peluko00) [reply stock_request_picking_type/tests/test_stock_picking_type.py]: Done!

File: stock_request_picking_type/tests/test_stock_request.py (hunk 1 above)
Comment (peluko00) [reply stock_request_picking_type/tests/test_stock_request.py]: Done!

File: stock_request_picking_type/tests/test_stock_request.py [hunk 3]
```diff
@@ -7,46 +7,11 @@
...
-        )
-
-    def _create_user(self, name, group_ids, company_ids):
+    @classmethod
+    def _create_user(cls, name, group_ids, company_ids):
         return (
-            self.env["res.users"]
-            .with_context({"no_reset_password": True})
+            cls.env["res.users"]
+            .with_context(**{"no_reset_password": True})
```
Comment (rousseldenis) [inline stock_request_picking_type/tests/test_stock_request.py]: You can avoid such writing by using arguments directly:
`.with_context(no_reset_password=True)`

File: stock_request_picking_type/tests/test_stock_request.py [hunk 4]
```diff
@@ -185,4 +190,4 @@ def test_create(self):
         order = self.request_order.with_user(self.stock_request_user).create(order_vals)
 
         # test create()
-        self.assertEqual(order.picking_type_id, new_pick_type)
+        self.assertNotEqual(order.picking_type_id, new_pick_type)
```
Comment (rousseldenis) [inline stock_request_picking_type/tests/test_stock_request.py]: Why is this changed ?

File: stock_request_picking_type/tests/test_stock_request.py (hunk 3 above)
Comment (peluko00) [reply stock_request_picking_type/tests/test_stock_request.py]: Fix it!

File: stock_request_picking_type/tests/test_stock_request.py (hunk 4 above)
Comment (peluko00) [reply stock_request_picking_type/tests/test_stock_request.py]: I remove the test because is not impacts  in the total coverage at the module and now is not possible to modify the warehouse and location after creation in form.

**Review by celm1990 — APPROVED** [review]
LGTM tested functionally

File: stock_request_picking_type/models/stock_request_order.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                ],
+                limit=1,
+            )
+            .id
+        )
+
+    picking_type_id = fields.Many2one(
+        comodel_name="stock.picking.type",
+        string="Operation Type",
+        default=_get_default_picking_type,
```
Comment (rousseldenis) [inline stock_request_picking_type/models/stock_request_order.py]: This deserves a lambda function instead.

File: stock_request_picking_type/models/stock_request_order.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+from odoo import api, fields, models
+
+
+class StockRequestOrder(models.Model):
+    _inherit = "stock.request.order"
+
+    @api.model
+    def _get_default_picking_type(self):
+        companies = self._context.get("allowed_company_ids", []).copy()
```
Comment (rousseldenis) [inline stock_request_picking_type/models/stock_request_order.py]: _context is deprecated since a long time: https://github.com/odoo/odoo/blob/16.0/odoo/models.py#L5160

File: stock_request_picking_type/models/stock_request_order.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        companies.append(False)
+        return (
+            self.env["stock.picking.type"]
+            .search(
+                [
+                    ("code", "=", "stock_request_order"),
+                    "|",
+                    ("warehouse_id.company_id", "in", companies),
```
Comment (rousseldenis) [inline stock_request_picking_type/models/stock_request_order.py]: Moreover, I think this is not necessary as there is a record rule on picking type. I'm wondering why it is not sufficent to search on...

File: stock_request_picking_type/models/stock_request_order.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+    picking_type_id = fields.Many2one(
+        comodel_name="stock.picking.type",
+        string="Operation Type",
+        default=_get_default_picking_type,
+        required=True,
+    )
+
+    @api.onchange("warehouse_id")
+    def onchange_warehouse_picking_id(self):
```
Comment (rousseldenis) [inline stock_request_picking_type/models/stock_request_order.py]: This should now be avoided. You can change the picking_type_id field to a computed/stored/readonly=False field instead and change this to _compute method.

File: stock_request_picking_type/models/stock_request_order.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                    ("code", "=", "stock_request_order"),
+                    ("warehouse_id", "=", self.warehouse_id.id),
+                ],
+                limit=1,
+            )
+            if picking_type_id:
+                self._origin.write({"picking_type_id": picking_type_id.id})
+
+    @api.model
+    def create(self, vals):
```
Comment (rousseldenis) [inline stock_request_picking_type/models/stock_request_order.py]: This can be avoided too referring the remark above.

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review

File: stock_request_picking_type/models/stock_request_order.py (hunk 5 above)
Comment (rousseldenis) [reply stock_request_picking_type/models/stock_request_order.py]: Update: No more default if changed to compute.

File: stock_request_picking_type/models/stock_request_order.py (hunk 6 above)
Comment (peluko00) [reply stock_request_picking_type/models/stock_request_order.py]: Still using here:
https://github.com/odoo/odoo/blob/17.0/odoo/models.py#L1158

Comment (rousseldenis) [reply stock_request_picking_type/models/stock_request_order.py]: That's not why they are using it, we should do it too :see_no_evil:

Comment (peluko00) [reply stock_request_picking_type/models/stock_request_order.py]: With `self.env.context` much better no?

Comment (rousseldenis) [reply stock_request_picking_type/models/stock_request_order.py]: Yes that's the one

**Comment by peluko00** [comment]
All improvements done @rousseldenis

File: stock_request_picking_type/models/stock_request_order.py [hunk 10]
```diff
@@ -7,43 +7,32 @@
...
-            .id
-        )
-
     picking_type_id = fields.Many2one(
         comodel_name="stock.picking.type",
         string="Operation Type",
-        default=_get_default_picking_type,
+        compute="_compute_picking_type_id",
         required=True,
+        store=True,
```
Comment (rousseldenis) [inline stock_request_picking_type/models/stock_request_order.py:15]: You should add readonly=False too.

File: stock_request_picking_type/models/stock_request_order.py (hunk 9 above)
Comment (rousseldenis) [reply stock_request_picking_type/models/stock_request_order.py]: create() override should be removed.

Comment (peluko00) [reply stock_request_picking_type/models/stock_request_order.py]: But the field is required

Comment (rousseldenis) [reply stock_request_picking_type/models/stock_request_order.py]: The compute will do the job as its aim is to get a default value.

Comment (peluko00) [reply stock_request_picking_type/models/stock_request_order.py]: The compute is not the same as default because on create can't assign a value

File: stock_request_picking_type/models/stock_request_order.py (hunk 10 above)
Comment (peluko00) [reply stock_request_picking_type/models/stock_request_order.py:15]: > You should add readonly=False too.

For the default for computed fields is readonly=True

File: stock_request_picking_type/models/stock_request_order.py (hunk 9 above)
Comment (rousseldenis) [reply stock_request_picking_type/models/stock_request_order.py]: Use precompute=True

Comment (peluko00) [reply stock_request_picking_type/models/stock_request_order.py]: Fix it!

File: stock_request_picking_type/models/stock_request_order.py (hunk 10 above)
Comment (peluko00) [reply stock_request_picking_type/models/stock_request_order.py:15]: Fix it!

**Comment by peluko00** [comment]
It's ready for merge @rousseldenis ?

**Comment by peluko00** [comment]
Any update @rousseldenis ? I need that for merge that [module](https://github.com/OCA/field-service/pull/1172)

**Review by rousseldenis — APPROVED** [review]
